sympy/solvers/tests/test_solvers.py

```diff
@@ -2712,3 +2712,15 @@ def test_solve_Piecewise():
         (46*x - 3*(x - 6)**2/2 - 276, (x >= 6) & (x < 10)),
         (0, x < 10),  # this will simplify away
         (S.NaN,True)))
+
+
+def test_issue_27854():
+    # https://github.com/sympy/sympy/issues/27854
+    m, n = symbols("m n")
+    A = MatrixSymbol("A", m, n)
+    x = MatrixSymbol("x", n, 1)
+    b = MatrixSymbol('b', m, 1)
+    r = A * x - b
+    f = r.T * r
+    grad_f = f.diff(x)
+    raises(ValueError, lambda: grad_f.as_explicit())
```

Comment:
This should be testing what happens with `solve`. Ideally find where there are any existing tests for `solve` with a `Matrix` as input and then add a test for this along with the other tests.